cirq/pasqal/pasqal_device.py

```diff
@@ -0,0 +1,208 @@
+
+        for qub in operation.qubits:
+            if not isinstance(qub, ThreeDGridQubit):
+                raise ValueError('{} is not a 3D grid qubit '
+                                 'for gate {!r}'.format(qub, operation.gate))
+
+        if isinstance(operation.gate,
+                      (cirq.ops.MeasurementGate, cirq.ops.IdentityGate)):
+            return
+
+        # Verify that a controlled gate operation is valid
+        if isinstance(operation, cirq.ops.GateOperation):
+            if len(operation.qubits) > self._max_parallel_c + \
+                    self._max_parallel_t:
+                raise ValueError("Too many qubits acted on in parallel by a"
+                                 "controlled gate operation")
+            if len(operation.qubits) > 1:
+                for p in operation.qubits:
+                    for q in operation.qubits:
+                        if self.distance(p, q) > self.control_radius:
+                            raise ValueError("Qubits {!r}, {!r} are too "
+                                             "far away".format(p, q))
+
+        # Verify that a valid number of Z gates are applied in parallel
+        if isinstance(operation.gate, cirq.ops.ZPowGate):
+            if len(operation.qubits) > self._max_parallel_z:
+                raise ValueError("Too many Z gates in parallel")
+
+        # Verify that a valid number of XY gates are applied in parallel
+        if isinstance(
+                operation.gate,
+            (cirq.ops.XPowGate, cirq.ops.YPowGate, cirq.ops.PhasedXPowGate)):
+            if (len(operation.qubits) > self._max_parallel_xy and
+                    len(operation.qubits) != len(self.qubit_list())):
+                raise ValueError("Bad number of X/Y gates in parallel")
+
+    def duration_of(self, operation: cirq.ops.Operation):
+        """
+        Provides the duration of the given operation on this device.
+
+        Args:
+            operation: the operation to get the duration of
+
+        Returns:
+            The duration of the given operation on this device
+
+        Raises:
+            ValueError: If the operation provided doesn't correspond to a native
+                gate
+        """
+        self.validate_operation(operation)
+        if isinstance(operation,
+                      (cirq.ops.GateOperation, cirq.ops.ParallelGateOperation)):
+            if isinstance(operation.gate, cirq.ops.MeasurementGate):
+                return self._measurement_duration
+        return self._gate_duration
+
+    def distance(self, p: 'cirq.Qid', q: 'cirq.Qid') -> float:
+        """
+        Returns the distance between two qubits.
+
+        Args:
+            p: qubit involved in the distance computation
+            q: qubit involved in the distance computation
+
+        Returns:
+            The distance between qubits p and q, in lattice spacing units.
+
+        """
+        if not isinstance(q, ThreeDGridQubit):
```

Comment:
In line 50 of this file, you have 
```python 
if not isinstance(q, ThreeDGridQubit):
                raise TypeError('Unsupported qubit type: {!r}'.format(q))
```
For the same exact code, you raise a `TypeError` there and a `ValueError` here. What's the difference?

Reply:
Now raising a TypeError everywhere.